README.md
# mc-bounty-processor
[![Latest Release](https://img.shields.io/github/v/release/Ubunfu/mc-bounty-processor)](https://github.com/Ubunfu/mc-bounty-processor/releases)
[![codecov](https://codecov.io/gh/Ubunfu/mc-bounty-processor/branch/master/graph/badge.svg?token=K1YOPWSEQD)](https://codecov.io/gh/Ubunfu/mc-bounty-processor)
[![CircleCI](https://img.shields.io/circleci/build/github/Ubunfu/mc-bounty-processor?logo=circleci)](https://app.circleci.com/pipelines/github/Ubunfu/mc-bounty-processor)
![Contrubutors](https://img.shields.io/github/contributors/Ubunfu/mc-bounty-processor?color=blue)
![Last Commit](https://img.shields.io/github/last-commit/Ubunfu/mc-bounty-processor)

Processes bounty requests submitted by [mc-bounty-scraper](https://github.com/Ubunfu/mc-bounty-scraper).  Valid bounties will be awarded by submitting a request to the [mc-wallet](https://github.com/Ubunfu/mc-wallet) service.

This service runs as an AWS lambda function.

## Configuration
### IAM Role
AWS's standard IAM role for Lambda micro services is plenty sufficient. The only access that is required is read/write for DynamoDB, and write to CloudWatch logs.

* AWSLambdaMicroserviceExecutionRole
* AWSLambdaBasicExecutionRole

### Environment Variables
| Parameter          | Description                                                                       | Default | Required? |
|--------------------|-----------------------------------------------------------------------------------|---------|-----------|
| TABLE_BOUNTIES     | The name of a DynamoDB table containing valid bounties and their values.          | n/a     | Yes       |
| SERVICE_WALLET_URL | The full URL of the mc-wallet payment service.                                    | n/a     | Yes       |
| LOGGER_ENABLED     | Boolean value controlling writing of logs. Useful to turn off for test execution. | n/a     | Yes       |
| DISCORD_WEBHOOK_URL | Discord webhook URL used to post notifications when bounties are turned in.      | n/a     | Yes       |
| DISCORD_WEBHOOK_THUMBNAIL_URL | URL of a thumbnail image to use for the Discord notification.          | n/a     | Yes       |

CHANGELOG.md
# Changelog
This project attempts to conform to [SemVer](https://semver.org/) guidelines whenever possible.

## [v1.1.0]
* [Adding 1.17 and 1.18 bounties](https://github.com/Ubunfu/mc-bounty-processor/pull/12)

## [v1.0.0]
* [Adding database to serverless stack](https://github.com/Ubunfu/mc-bounty-processor/pull/9)
* [Configuring Discord cicd notifications](https://github.com/Ubunfu/mc-bounty-processor/pull/10)

## [v0.3.0]
* [Adopting serverless framework](https://github.com/Ubunfu/mc-bounty-processor/pull/7)

## [v0.2.0]
* [Fixing wallet service interface](https://github.com/Ubunfu/mc-bounty-processor/pull/3)
* [Adding bounty data load files](https://github.com/Ubunfu/mc-bounty-processor/pull/4)
* [Adding a Discord notification when a bounty is turned in](https://github.com/Ubunfu/mc-bounty-processor/pull/5)

## [v0.1.0]
* Initial implementation

test/indexTest.js
const index = require('../src/index.js');
const expect = require('chai').expect;
const sinon = require('sinon');
const dbService = require('../src/db/dbService.js');
const walletService = require('../src/wallet/walletService.js');
const axios = require('axios');

const TEST_PLAYER = "whooshbokoo";
const TEST_BOUNTY = "Diamonds!";
const TEST_BOUNTY_VALUE = 99;
const TEST_DISCORD_WEBHOOK_PAYLOAD = {
    "embeds": [
        {
            "author": {
                "name": "Bounty Hunter"
            },
            "thumbnail": {
                "url": process.env.DISCORD_WEBHOOK_THUMBNAIL_URL
            },
            "title": `Bounty turned in by ${TEST_PLAYER}`,
            "color": 8421504,
            "fields": [
                {
                    "name": TEST_BOUNTY,
                    "value": TEST_BOUNTY_VALUE
                }
            ]
        }
    ]
};

describe('index: When receives bounty request event', function() {
    describe('And DB returns a matching bounty', function() {
        let dbMock = null;
        beforeEach(function() {
            dbMock = sinon.stub(dbService, "getBounty")
                .returns({
                    Item: {
                        BountyId: TEST_BOUNTY,
                        Value: TEST_BOUNTY_VALUE
                    }
                });
        });
        afterEach(function() {
            dbMock.restore();
        });

        describe('And wallet-service returns 200', function() {
            let walletMock, handlerResponse, webHookMock = null;
            beforeEach(async function() {
                walletMock = sinon.stub(walletService, "pay")
                    .returns({
                        statusCode: '200'
                    });
                webHookMock = sinon.stub(axios, "post").returns(null);
                handlerResponse = await index.handler({
                    body: `{"player": "${TEST_PLAYER}", "achievement": "${TEST_BOUNTY}"}`
                });
            });
            afterEach(function() {
               walletMock.restore();
               webHookMock.restore();
            });

            it('Should return 200 status code', function() {
                expect(handlerResponse.statusCode).to.be.equal('200');
            });

            it('Should pay player', function() {
                expect(walletMock.calledOnceWith(`${TEST_PLAYER}`, TEST_BOUNTY_VALUE)).to.be.true;
            });

            it('Should post notification to Discord', function() {
                expect(webHookMock.calledOnce).to.be.true;
                expect(webHookMock.lastCall.args[1]).to.be.deep.equal(TEST_DISCORD_WEBHOOK_PAYLOAD);
            });
        });

        describe('And wallet-service returns error', function() {
            let walletMock, handlerResponse = null;
            beforeEach(async function() {
                walletMock = sinon.stub(walletService, "pay")
                    .throws({
                        response: {
                            statusCode: '500'
                        }
                    });
                handlerResponse = await index.handler({
                    body: `{"player": "${TEST_PLAYER}", "achievment": "${TEST_BOUNTY}"}`
                });
            });
            afterEach(function() {
               walletMock.restore();
            });

            it('Should return 500 status code', function() {
                expect(handlerResponse.statusCode).to.be.equal('500');
            });
            
            it('Should return error message', function() {
                let errorResp = JSON.parse(handlerResponse.body);
                expect(errorResp.error).to.be.equal("error posting bounty payment");
            });
        });
    });

    describe('And DB does not find a matching bounty', function() {
        let dbMock, walletSpy, handlerResponse = null;
        beforeEach(async function() {
            dbMock = sinon.stub(dbService, "getBounty")
                .returns({});
            walletSpy = sinon.spy(walletService, "pay");
            handlerResponse = await index.handler({
                body: `{"player": "${TEST_PLAYER}", "achievment": "${TEST_BOUNTY}"}`
            });
        });
        afterEach(function() {
            dbMock.restore();
            walletSpy.restore();
        });

        it('Should return 400', function() {
            expect(handlerResponse.statusCode).to.be.equal('400');
        });

        it('Should return error message', function() {
            let errorResp = JSON.parse(handlerResponse.body);
            expect(errorResp.error).to.be.equal("invalid bounty");
        });

        it('Should not pay player', function() {
            expect(walletSpy.callCount).to.be.equal(0);
        });
    });
});

test/db/dbServiceTest.js
const sinon = require('sinon');
const dbService = require('../../src/db/dbService.js');
const expect = require('chai').expect;

describe('dbService: When database returns a record', function() {
    let dbServiceResp = null;
    beforeEach(async function() {
        const mockDocClient = {
            get: sinon.stub().returnsThis(),
            promise: sinon.stub().resolves({
                Item: {
                    BountyId: 'an item', 
                    Value: 5
                }})
        };
        dbServiceResp = await dbService.getBounty(mockDocClient, "an item");
    });

    it('Should return the same record', function() {
        expect(dbServiceResp.Item.Value).to.be.equal(5);
    });
});

describe('dbService: When database returns an error', function() {
    let dbServiceResp = null;
    beforeEach(async function() {
        const mockDocClient = {
            get: sinon.stub().throwsException("error")
        }
        dbServiceResp = await dbService.getBounty(mockDocClient, "an item");
    });

    it('Should return the error', function() {
        expect(dbServiceResp.name).to.be.equal("error");
    })
});

test/wallet/walletServiceTest.js
const walletService = require('../../src/wallet/walletService.js');
const axios = require('axios');
const expect = require('chai').expect;
const sinon = require('sinon');

describe('walletService: When wallet service call returns HTTP 200', function() {
    let walletResp, axiosMock = null;
    beforeEach(async function() {
        axiosMock = sinon.stub(axios, "post")
            .returns({
                status: 200,
                data: 'walletResponse'
            });
        walletResp = await walletService.pay("whooshbokoo", 100);
    });
    afterEach(function() {
        axiosMock.restore();
    });

    it('Should return response data', function() {
        expect(walletResp.data).to.be.equal('walletResponse')
    });
});

describe('walletService: When wallet service returns HTTP error', function() {
    let walletResp, axiosMock = null;
    beforeEach(function() {
        axiosMock = sinon.stub(axios, "post")
            .throws({response: {status: 500}});
    });
    afterEach(function() {
        axiosMock.restore();
    });
    
    it('Should return error', async function() {
        walletResp = await walletService.pay("whooshbokoo", 10);
        expect(walletResp.response.status).to.be.equal(500);
    });
});

resources/db/makeBatch.js
const fs = require('fs');

const TABLE_NAME = process.env.TABLE_NAME
const SCAN_OUTPUT_FILE = process.env.SCAN_OUTPUT_FILE
const BATCH_OUTPUT_FILE_PREFIX = 'bounties'

async function makeBatch() {
    const backupString = await fs.readFileSync(SCAN_OUTPUT_FILE, 'utf-8')
    const backupBounties = JSON.parse(backupString)

    let batchRequests = []
    backupBounties.forEach(backupBounty => {
        batchRequests.push({
            PutRequest: {
                Item: decorateWithDataTypes(backupBounty)
            }
        })
    })
    
    for (let batchFileIndex = 0; batchFileIndex < (batchRequests.length/25); batchFileIndex++) {
        let batchObject = {}
        batchObject[TABLE_NAME] = []
    
        for (let requestIndex = 0; requestIndex < 25; requestIndex++) {
            const overallRequestIndex = (batchFileIndex*25+requestIndex)
            if (overallRequestIndex == batchRequests.length) {
                break
            }
            batchObject[TABLE_NAME].push(batchRequests[overallRequestIndex])
        }
        await fs.writeFileSync(BATCH_OUTPUT_FILE_PREFIX+`_${batchFileIndex}.json`, JSON.stringify(batchObject), 'utf8')
    }
}

function decorateWithDataTypes(bounty) {
    // {"itemName":"Cobblestone","itemId":"minecraft:cobblestone","price":1,"sellPrice":0.05}
    bounty.BountyId = {
        S: bounty.BountyId
    }
    bounty.Value = {
        N: `${bounty.Value}`
    }
    return bounty
}

makeBatch()

src/index.js
require('dotenv').config();
const dbService = require('./db/dbService.js');
const { log } = require('./util/logger.js');
const walletService = require('./wallet/walletService.js');
const AWS = require ('aws-sdk');
const axios = require('axios');

exports.handler = async (event, context) => {
    await log('Received event: ' + JSON.stringify(event, null, 2));

    let body;
    let statusCode = '200';
    const headers = {
        'Content-Type': 'application/json',
    };

    const bounty = JSON.parse(event.body).achievement;
    const player = JSON.parse(event.body).player;
    const docClient = new AWS.DynamoDB.DocumentClient();
    var bountyValue;
    
    try {
        let bountyRecord = await dbService.getBounty(docClient, bounty);
        await log("db returned: " + JSON.stringify(bountyRecord));
        
        bountyValue = await bountyRecord.Item.Value;
        await log("bounty is: " + bountyRecord.Item.Value);
    } catch (err) {
        statusCode = '400';
        body = {
            error: "invalid bounty",
            errorDetail: err.message
        };
    }

    try {
        if (statusCode == '200') {
            await walletService.pay(player, bountyValue);
            try {
                await axios.post(
                    process.env.DISCORD_WEBHOOK_URL,
                    await buildDiscordPayload(player, bounty, bountyValue)
                );
            } catch (err) {
                log('Error sending notification to Discord: ' + err);
            }
        }
    } catch (err) {
        log(err);
        statusCode = '500';
        body = {
            error: "error posting bounty payment",
            errorDetail: `${err.response}`
        };
    }

    body = JSON.stringify(body);
    return {
        statusCode,
        body,
        headers,
    };
};

async function buildDiscordPayload(player, bounty, bountyValue) {
    return {
        "embeds": [
            {
                "author": {
                    "name": "Bounty Hunter"
                },
                "thumbnail": {
                    "url": process.env.DISCORD_WEBHOOK_THUMBNAIL_URL
                },
                "title": `Bounty turned in by ${player}`,
                "color": 8421504,
                "fields": [
                    {
                        "name": `${bounty}`,
                        "value": bountyValue
                    }
                ]
            }
        ]
    }
}